presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -217,7 +224,7 @@ private SerializedPage postProcessPage(SerializedPage page)
     public boolean isFinished()
     {
         throwIfFailed();
-        return isClosed() && completedClients.size() == locations.size();
+        return isClosed() && completedClients.size() == allClients.size();
```

Comment:
1. Explain the guarantee needed here in a comment.
2. Edit `addLocation` to satisfy that guarantee.